### DISCUSSION TITLE
Examples for Mozzi classes

### DISCUSSION THREAD

**Discussion by jaapnoordzij** [pschatzmann/arduino-audio-tools#1443] [Q&A] [answered]
Hi Phil and all others

I was experimenting with Mozzi and so I was happy to see the new (?) Mozzi classes. Since I am not very experienced I tried to combine some basis examples from audio-tools sending a stream to I2S with mozzi generating a sineWave. However the sketch below compiles with errors like:
- WifiClientSecure not defined (I can correct that)
- Mozzi.h not defined in AudioLibs/AudioMozzi.h  - I changed that from Mozzi.h to MozziGuts.h but that created only more errors.

I compile this for an Nano ESP32 (S3) and my libdeps include
  https://github.com/pschatzmann/arduino-audio-tools
  https://github.com/pschatzmann/Mozzi

Could you give me a simple hint on what is wrong ?  
I am in no hurry and I am happy to wait until you can find the time to include some real Mozzi Examples for audio-tools

```
/**
 * @file 
 * @author
 * @brief 
 * @copyright
 */
#include <WiFiClientSecure.h>

#include <AudioTools.h>
#include "AudioLibs/AudioMozzi.h"

#include <Oscil.h>
#include <tables/sin2048_int8.h>

#define CONTROL_RATE 64 // Hz, powers of 2 are most reliable

Oscil <SIN2048_NUM_CELLS, AUDIO_RATE> aSin(SIN2048_DATA);
byte gain = 255;

AudioInfo info(44100, 2, 16);
I2SStream i2s; 

// Setup Mozzi generator
MozziGenerator<int16_t> mozziWave(); // Subclass of SoundGenerator
MozziStream<int16_t> msound(mozziWave); // Stream generated from Mozzi

VolumeStream vol(i2s);
StreamCopy copier(vol, msound);   // copies sound/music into i2s through vol

void updateControl () {
  // as byte, this will automatically roll around to 255 when it passes 0
  gain = gain - 3 ;

}

int updateAudio () {
  return (aSin.next()* gain)>>8; // shift back to STANDARD audio range, like /256 but faster
}


// Arduino Setup
void setup(void) {  
  // Open Serial 
  Serial.begin(115200);
  delay(1000);
  Serial.println(F("Basic experimenating audio_tools"));
  AudioLogger::instance().begin(Serial, AudioLogger::Info);

  aSin.setFreq(3320); // set the oscillator frequency

  // start I2S
  auto config = i2s.defaultConfig(TX_MODE);
  config.pin_bck = 10;  // Nano D7
  config.pin_ws = 17;   // Nano D8
  config.pin_data = 21; // Nano D10
  config.copyFrom(info); 
  i2s.begin(config);

  // setup volume
  auto config_vol = vol.defaultConfig();
  config_vol.copyFrom(info);
  config_vol.volume = 0.4;  // half the volume
  vol.begin(config_vol);

  // Setup Mozzi
  mozziWave.begin(info);
}

// Arduino loop - copy sound to out 
void loop() {
  copier.copy();
}
```

**Comment by pschatzmann** [comment, marked as answer]
Actually the MozziStream has been retired and does not work any more.
I have deleted it now in order to avoid any confusion.

I recommend to use the original Mozzi library implementation and use my Fork only if you need data input for the ESP32

**Comment by pschatzmann** [comment]
I gave this topic another thought and I have updated the code so that it works with the official Mozzi library.
Here is an [example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-dsp/examples-mozzi/control_gain/control_gain.ino)

**Reply by jaapnoordzij** [reply]
Awesome, thanks a lot Phil !
